### DISCUSSION TITLE
Heyu 2.10.3 (also: how to resolve issues with file permissions / group ownership of device file and/or lock directory -- any Heyu version)

### DISCUSSION THREAD

**Discussion by jkrzyszt** [HeyuX10Automation/heyu#36] [Q&A] [answered]
#### List of changes since release 2.10.2:

Fredrik Fornwall (1):
      Replace rindex(3) with strrchr(3)

Janusz Krzysztofik (2):
      Fix gcc 10+ build errors
      Heyu 2.10.3

#### [Changelog](http://github.com/HeyuX10Automation/heyu/compare/v2.10.2...v2.10.3) against last stable release 2.10.2

<hr /><em>This discussion was created from the release <a href='https://github.com/HeyuX10Automation/heyu/releases/tag/v2.10.3'>Heyu 2.10.3</a>.</em>

**Comment by npaisnel** [comment]
I have just come back to this to rebuild a new Raspberry Pi box to run some x-10 stuff .

i see my old instruction i crearted for my self points to 2.11RC3.
But looking here to see if that was the latest, I see that this 2.10.3 seems to be the last release.

I know very little (nothing about all of this)   more than following tutorials on how to do stuff.

Which should I be using 
I have just Ssh'd to my currently running box and i see in a src folder that I have been running v 2.11 RC3 since ..2018 maybe ?
Any advice appreciated. 

many thanks

**Comment by jkrzyszt** [comment]
Hi,

2.10.3 is the latest stable release.  Almost all fixes against 2.10.2 have also been applied to the master branch which is based on 2.11.rc3.  There is also a branch named topic/locking which contains one more fix supposed to resolve issues with stale lock files.  If possible, please download [a snapshot of that branch](https://github.com/HeyuX10Automation/heyu/archive/refs/heads/topic/locking.zip) and give it a try -- I need some feedback before I can decide if it may be merged to master as is and possibly submitted as a new release, or it needs more work.  If you have issues with it, please report and use [a snapshot of the master branch](https://github.com/HeyuX10Automation/heyu/archive/refs/heads/master.zip).

Thanks,
Janusz

**Reply by npaisnel** [reply]
I would love to help you with that ..but I am very much afraid i do not think I'd know what to do with those files.  I am just a very basic user that follows tutorials to get stuff working.

I have not touched my working system since I built it back in 2018.  using 11 rc3   

i am building up a new system now ... so could create a separate SD card build to try this out ..if

**Comment by jkrzyszt** [comment]
There is nothing special with sapshots from git branches.  Those are just ZIP archives of patched Heyu soruces.  Unpack one and follow standard installation instructions (see README file for details).

Thanks,
Janusz

**Reply by npaisnel** [reply]
I will try ..as I appreciate your work...
but remember here , with me you are dealing with some one that if I stick a USB data stick in my headless CLI only Raspebeeryy Pi, and I want to read the contents of the USB stick..I have to first Google even how to mount and access the USB stick to copy / read data off it.

I am taking a Pi with me to work today .  
The pi has my work image on it with 2.11RC3. 

I also have the with the files you link to on USB..so i'll see if I can copy them over and do the  make/ make is install stuff and get it to work

**Comment by npaisnel** [comment]
i am trying today
not wanting to start the device at the moment ..or anyheyu commands.  get a lock file error on heyu info

 heyu info
Couldn't remove stale lock file /var/lock/LCK..LCK..

..go past that 
stoped service..disalbed it 

now "heyu info " just stops everything can leave it 2-3 minutes..just stops with blinking cursor 

Will do another clean rebuild of Raspberry OS on to a fresh SD card and try again ..but I do have other jobs to do today

**Reply by npaisnel** [reply]
Fresh RPI Lite now and trying with "sudo make install

but no go with the Stale-lock file branch

**Comment by npaisnel** [comment]
OK, so each attempt I started with a Fresh 'install of Raspbian Stretch Lite 


if i use

 https://github.com/HeyuX10Automation/heyu/archive/v2.11-rc3.tar.gz

it works ..heyu drives the CM11

I can issue commands like heyu on a6 and the X-10 device responds


if I use 
https://github.com/HeyuX10Automation/heyu/archive/refs/tags/v2.10.3.tar.gz

then there is no Configure.sh file and the command  

sh ./Configure.sh  does not work


If I use

https://github.com/HeyuX10Automation/heyu/archive/refs/heads/topic/locking.zip

then everything goes this same as it does for 11rc3...until the command "heyu info"

Then the cursor just blinks and nothing further


I will try with the master version you linked to earlier this afternoon.


but so far.  on an RPi2 with Stretch Lite 
only version 11 RC3 works 


Yes master version works and 11 RC3 works ..

no others work for me

**Reply by npaisnel** [reply]
OK, so done using the clean install again using the snapshot of the master branch ..and that works fine 

But the other link 

https://github.com/HeyuX10Automation/heyu/archive/refs/heads/topic/locking.zip


does not work on a Raspberry Pi2 swith stretch and the supplied CM11 adapter cable.

I do not have a serial to USB FTDI cabe to tests only two Prolific cables.

**Reply by jkrzyszt** [reply]
@npaisnel thanks for testing and reporting.

Whenever you try 2.10.x, please use ./Configure, without the .sh extension.   It was renamed to ./Congifure.sh in 2.11 to avoid conflicts with autoconf's ./configure on case insensitive filesystems (that name change was reflected in README).

Regarding an issue with the topic/locking branch, please, if still possible, try to get verbose output by runninng 'heyu -v info' and report.

Thanks,
Janusz

**Reply by npaisnel** [reply]
I will do that .

i will start from a clean RPI Lite image and give another try 

2.11RC3 certainly works with the .sh extension.



My Home Vision device has now gone and died..serial chip seems to be FUBAR..so got that to deal with too..aaargggghhh  all at once...


https://groups.google.com/g/homevision-users/c/Yju6oXiRrsc

**Comment by npaisnel** [comment]
the locking branch

on a Raspberry Pi 2
clean write to an sd card from an old image i had created of Stretch lite, with just SSH and Samba ready configured.
 This version
https://github.com/HeyuX10Automation/heyu/archive/refs/heads/topic/locking.zip

I get ..as below..it jsut gets stuck in that loop until I Ctrl-C out of it


pi@litepi:~/src/heyu-topic-locking $ heyu -v info
Version:2.11-rc3
Searching for '/home/pi/.heyu/x10config'
Searching for '/etc/heyu/x10.conf'
Found configuration file '/etc/heyu/x10.conf'
Heyu directory /etc/heyu/ is writable.
Reading Heyu configuration file '/etc/heyu/x10.conf'
Trying to lock (/var/lock/LCK..heyu.write.ttyUSB0)
lockpid: Checking for file '/var/lock/LCK..LCK..'
lockpid: Checking for file '/var/lock/LCK..LCK..'
lockpid: Checking for file '/var/lock/LCK..LCK..'
lockpid: Checking for file '/var/lock/LCK..LCK..'
lockpid: Checking for file '/var/lock/LCK..LCK..'
lockpid: Checking for file '/var/lock/LCK..LCK..'
lockpid: Checking for file '/var/lock/LCK..LCK..'
lockpid: Checking for file '/var/lock/LCK..LCK..'
lockpid: Checking for file '/var/lock/LCK..LCK..'
lockpid: Checking for file '/var/lock/LCK..LCK..'
lockpid: Checking for file '/var/lock/LCK..LCK..'

**Reply by npaisnel** [reply]
I deleted the /etc/heyu directory and then did the master version you linked to 


root@litepi:/home/pi/src/heyu-master# heyu -v info
Version:2.11-rc3
Searching for '/root/.heyu/x10config'
Searching for '/etc/heyu/x10.conf'
Found configuration file '/etc/heyu/x10.conf'
Heyu directory /etc/heyu/ is writable.
Reading Heyu configuration file '/etc/heyu/x10.conf'
lockpid: Checking for file '/var/lock/LCK..heyu.write.ttyUSB0'
Trying to lock (/var/lock/LCK..heyu.write.ttyUSB0)
lockpid: Checking for file '/var/lock/LCK..heyu.write.ttyUSB0'
/var/lock/LCK..heyu.write.ttyUSB0 is locked
Alert acks are 0x52, 0x12, 0x42
lockpid: Checking for file '/var/lock/LCK..heyu.relay.ttyUSB0'
lockpid: Checking for file '/var/lock/LCK..heyu.relay.ttyUSB0'
There was already a relay running (pid = 7820)
xwrite() called, count=1
xread() called, count=14, timeout = 3
xread() returning 14 byte(s). The first is 0xff
xwrite() called, count=1
xread() called, count=1, timeout = 1
xread() returning 1 byte(s). The first is 0xeb
Checksum confirmed
xwrite() called, count=1
xread() called, count=1, timeout = 1
xread() returning 1 byte(s). The first is 0x55
Interface is ready.
The battery usage is unknown.
09:36:56 is the interface time
144 is the julian date
Tue ( 4) is the interface day of the week
Heyu version 2.11-rc3
Configuration at /etc/heyu/x10.conf
Powerline interface on /dev/ttyUSB0
Firmware revision Level = 8
Interface battery usage = Unknown
Raw interface clock: Tue, Day 144, 09:36:56
Searching for /etc/heyu/x10record
File /etc/heyu/x10record is absent
(--> Civil Time: Tue 25 May 2021   10:36:56 BST)
No schedule has been uploaded by Heyu.
Housecode = A
0 = off, 1 = on,               unit  16.......8...4..1
Last addressed device =       0x0002 (0000000000010000)
Status of monitored devices = 0x0840 (0000100000000001)
Status of dimmed devices =    0x0000 (0000000000000000)
lockpid: Checking for file '/var/lock/LCK..heyu.write.ttyUSB0'
munlock: Unlink file '/var/lock/LCK..heyu.write.ttyUSB0'
root@litepi:/home/pi/src/heyu-master#

**Comment by npaisnel** [comment]
2.10.3....



** Now become root and run 'make install' **

pi@litepi:~/src/heyu-2.10.3 $ sudo su
root@litepi:/home/pi/src/heyu-2.10.3# make install
mkdir -p -m 755 /usr/local/bin
cp heyu /usr/local/bin
cp: cannot create regular file '/usr/local/bin/heyu': Text file busy
Makefile:111: recipe for target '/usr/local/bin/heyu' failed
make: *** [/usr/local/bin/heyu] Error 1
root@litepi:/home/pi/src/heyu-2.10.3# sudo reboot now





pi@192.168.1.71's password: 
Linux litepi 4.14.79-v7+ #1159 SMP Sun Nov 4 17:50:20 GMT 2018 armv7l

The programs included with the Debian GNU/Linux system are free software;
the exact distribution terms for each program are described in the
individual files in /usr/share/doc/*/copyright.

Debian GNU/Linux comes with ABSOLUTELY NO WARRANTY, to the extent
permitted by applicable law.
Last login: Tue May 25 11:01:18 2021
pi@litepi:~ $ sudo su
root@litepi:/home/pi# cd src
root@litepi:/home/pi/src# ls
heyu-2.10.3  v2.10.3.tar.gz
root@litepi:/home/pi/src# cd heyu-2.10.3

root@litepi:/home/pi/src/heyu-2.10.3# make install
mkdir -p -m 755 /usr/local/bin
cp heyu /usr/local/bin
chgrp root  /usr/local/bin/heyu
chmod 755 /usr/local/bin/heyu
chown root /usr/local/bin/heyu
./install.sh

I did not find a Heyu configuration file.
Where would you like the sample Heyu configuration file installed?
  1. In directory /home/pi/.heyu/
  2. In subdirectory .heyu/ under a different user home directory
  3. In directory /etc/heyu  (for system-wide access)
  4. No thanks, I'll take care of it myself
Choice [1, 2, 3, 4] ? 3
Creating directory /etc/heyu with permissions rwxrwxrwx.
Adjust ownership and permissions as required.
The sample configuration file will be installed as /etc/heyu/x10.conf

I will add the TTY port for your CM11 to the config file
Specify /dev/ttyS0, /dev/ttyS1, etc., or the word dummy
To which port is the CM11 attached? /dev/ttyUSB0
Changing TTY permissions to 777
The permissions for the SPOOL directory (/var/tmp/heyu) are OK
The permissions for the directory /var/lock were set to 1777
cat install.sh >install 
chmod a+x install
root@litepi:/home/pi/src/heyu-2.10.3# 
root@litepi:/home/pi/src/heyu-2.10.3# heyu -v info
Version:2.10.3
Searching for '/root/.heyu/x10config'
Searching for '/etc/heyu/x10.conf'
Found configuration file '/etc/heyu/x10.conf'
Heyu directory /etc/heyu/ is writable.
Reading Heyu configuration file '/etc/heyu/x10.conf'
lockpid: Checking for file '/var/lock/LCK..heyu.write.ttyUSB0'
Trying to lock (/var/lock/LCK..heyu.write.ttyUSB0)
lockpid: Checking for file '/var/lock/LCK..heyu.write.ttyUSB0'
/var/lock/LCK..heyu.write.ttyUSB0 is locked
Alert acks are 0x52, 0x12, 0x42
lockpid: Checking for file '/var/lock/LCK..heyu.relay.ttyUSB0'
Relay lockfile not found - spawning heyu_relay.
starting heyu_relay
xwrite() called, count=1
xread() called, count=14, timeout = 3
xread() returning 14 byte(s). The first is 0xff
xwrite() called, count=1
xread() called, count=1, timeout = 1
xread() returning 1 byte(s). The first is 0xeb
Checksum confirmed
xwrite() called, count=1
xread() called, count=1, timeout = 1
xread() returning 1 byte(s). The first is 0x55
Interface is ready.
The battery usage is unknown.
10:05:18 is the interface time
144 is the julian date
Tue ( 4) is the interface day of the week
Heyu version 2.10.3
Configuration at /etc/heyu/x10.conf
Powerline interface on /dev/ttyUSB0
Firmware revision Level = 8
Interface battery usage = Unknown
Raw interface clock: Tue, Day 144, 10:05:18
Searching for /etc/heyu/x10record
File /etc/heyu/x10record is absent
(--> Civil Time: Tue 25 May 2021   11:05:18 BST)
No schedule has been uploaded by Heyu.
Housecode = A
0 = off, 1 = on,               unit  16.......8...4..1
Last addressed device =       0x0002 (0000000000010000)
Status of monitored devices = 0x0840 (0000100000000001)
Status of dimmed devices =    0x0000 (0000000000000000)
lockpid: Checking for file '/var/lock/LCK..heyu.write.ttyUSB0'
munlock: Unlink file '/var/lock/LCK..heyu.write.ttyUSB0'
root@litepi:/home/pi/src/heyu-2.10.3#

**Comment by jkrzyszt** [comment]
@npaisnel thanks a lot for testing and reporting.  You effort has helped me to identify a bug in the topic/locking branch: lockpid function is using a malformed name of the lock file. I'll take a closer look at the code and submit a fix.

Regarding the issue with make install, there is a recommendation to stop running heyu processes before an update / downgrade attempt.  With that in mind, both 2.10.3 and master snapshot seem working correctly for you.

Thanks,
Janusz

**Reply by npaisnel** [reply]
yes , I guess the heyu process may have been running  the first time i tried it, but not the second.

Now I know what to do i can test the new lock branch  again when needed

**Comment by ghutib** [comment]
Thanks for continuing the heyu development!

I recently installed heyu 2.10.3 on a Centos 7 system running Asterisk and have been successfully executing heyu commands from the Asterisk dialplan (as user asterisk group asterisk) until today, when the system was shut off abruptly by a power failure.

Now the commands from Asterisk fail with a message in /var/log/messages "heyu: The lock directory /var/lock is not writable." I can still execute heyu commands as root. I've power cycled and rebooted but neither fixed the problem. Here's /var/run/lock (which /var/lock points to):

drwx------ 2 root root  60 Jun 15 22:30 iscsi
-rw-r--r-- 1 root root   0 Jun 15 22:30 kdump
-rw-r--r-- 1 root root  11 Jun 15 22:30 LCK..
-rw-rw-rw- 1 root root   9 Jun 15 22:38 LCK..heyu.relay.ttyS0
-rw-rw-rw- 1 root root   9 Jun 15 22:38 LCK..ttyS0
drwxrwxr-x 2 root lock  40 Jun 15 22:32 lockdev
drwx------ 2 root root  40 Jun 15 22:30 lvm
drwxr-xr-x 2 root root 180 Jun 15 22:32 subsys
drwxr-xr-x 2 uucp uucp  40 Jun 15 22:30 uucp
drwxr-xr-x 2 root root  60 Jun 15 22:30 vdo
-rw-r--r-- 1 root root   0 Jun 15 22:30 vdo-config-singletons

The Asterisk dialplan runs heyu.sh (owned by asterisk) which consists of:

#!/bin/bash
/usr/local/bin/heyu $1 $2

Before I start changing things blindly, I thought I'd better ask the experts.

Thanks.

--ghutib

**Comment by jkrzyszt** [comment]
The problem with insufficient permissions to /var/lock is distribution specific, beyond the scope of Heyu source code.
Quoting https://forums.centos.org/viewtopic.php?t=65936 ("Permission problem when using lock files"):
"The solution to this is that your app should ship a /etc/tmpfiles.d config file that tells systemd-tmpfiles to create a /run/lock/yourapp directory, owned by your app user. So, don't create your lockfile directly under /run/lock but in a subdorectory off that that is created by systemd-tmpfiles for you at boot time (or when `systemd-tmpfiles --create /etc/tmpfiles.d/yourapp.conf` is run)."
You can use this method to either change group ownership and group permissions of /var/run/lock, or use a dedicated heyu subdir as suggested and build heyu with -DLOCKDIR=/var/lock/heyu added to CFLAGS.  If the latter then either edit Heyu ./Configure script and re-run it or edit Makefile created by ./Configure, then rebuild Heyu with make.
Thanks,
Janusz

**Comment by ghutib** [comment, marked as answer]
Hi Janusz,

Thanks—that worked! I edited /usr/lib/tmpfiles.d/legacy.conf and changed the line with:

d /run/lock 0755 root root -

to:

d /run/lock 0777 root root -

Now I have another problem. You may remember that I’m running Heyu on my Asterisk box with the dialplan invoking Heyu.

When invoking Heyu from Asterisk (user asterisk), it fails due to the permissions on /dev/ttyS0 being:

crw-rw---- 1 root dialout 4, 64 Jul 16 21:02 /dev/ttyS0

First, I tried adding asterisk to the dialout group, but that didn’t work.

Then I tried changing the ttyS0 permissions from:

crw-rw---- 1 root dialout 4, 64 Jul 16 21:02 /dev/ttyS0

to:

crw-rw-rw- 1 root dialout 4, 64 Jul 16 21:02 /dev/ttyS0

…and that worked.

But then intermittently, somebody is changing the permissions for ttyS0 back to:

crw-rw---- 1 root dialout 4, 64 Jul 16 21:02 /dev/ttyS0

Is there a better, persistent fix?

Thanks.

Regards,
Steve

**Reply by jkrzyszt** [reply]
On modern distributions udev is used to maintain ownership and permissions of device files.  Read 'man udev' to identifiy directories where udev rule files are stored, find one with rules for serial devices, place a copy of that file in a udev rules directory under /etc and modify it as needed.

**Comment by ghutib** [comment]
Hi Janusz,

Thanks for pointing me to udev. I haven’t been able to figure out where udev on CentOS 7 is told to set the mode for /dev/ttyS0 to 660, although udevadm says that it does:

udevadm test /devices/pnp0/00:06/tty/ttyS0 gives this (partial) output:

GROUP 18 /usr/lib/udev/rules.d/50-udev-default.rules:22
no db file to read /run/udev/data/+pnp:00:06: No such file or directory
handling device node '/dev/ttyS0', devnum=c4:64, mode=0660, uid=0, gid=18
preserve permissions /dev/ttyS0, 020660, uid=0, gid=18
preserve already existing symlink '/dev/char/4:64' to '../ttyS0'
created db file '/run/udev/data/c4:64' for '/devices/pnp0/00:06/tty/ttyS0'

What am I missing?

Thanks.

Regards,
Steve

**Reply by jkrzyszt** [reply]
From the above I conclude file permissions and group ownership of /dev/ttyS0 are defined in /usr/lib/udev/rules.d/50-udev-default.rules line 22 on your system

**Comment by ghutib** [comment]
Hi Janusz,

Here’s /usr/lib/udev/rules.d/50-udev-default.rules line 22:

KERNEL=="tty[A-Z]*[0-9]|pppox[0-9]*|ircomm[0-9]*|noz[0-9]*|rfcomm[0-9]*", GROUP="dialout"

I’m also attaching 50-udev-default.rules.

Regards,
Steve

**Reply by jkrzyszt** [reply]
Have you followed my advice, copied that file to udev rules subdir under /etc and modified that line as needed?  E.g., add something like MODE="0666" to it, or change GROUP="dialout" to something else?  Instead of copying the whole file and modifying it, you can also create a new file, best with a name starting with 99-z to make sure it will be parsed last and override other rules, with a rule like that for ttyS0 specifically.  Changes to /dev/ttyS0 should be visible after reboot.

**Comment by ghutib** [comment]
Hi Janusz,

You’re right, I should have tried that first--it worked.

I was focused on figuring out where the /dev/ttyS0 permissions had been set to 0660 in the first place.

Thank you!

Regards,
Steve

**Comment by jkrzyszt** [comment]
Steve and All,
If you have a question not directly related to the topic of a thread, please open a new thread instead of appending your message under an unrelated subject.  This way information will be more easy to find for others.
Thanks,
Janusz
